pootle/apps/pootle_translationproject/views.py

```diff
@@ -43,120 +57,323 @@ def admin_permissions(request, translation_project):
+                                if k in [
+                                    "language_code",
+                                    "project_code"]}))
+                except TranslationProject.DoesNotExist:
+                    pass
+
+            # if we get here - the TP does not exist
+            user_choice = self.request.COOKIES.get(
+                'user-choice', None)
+            if user_choice:
+                url = None
+                if user_choice == 'language':
+                    url = reverse(
+                        'pootle-language-browse',
+                        args=[kwargs["language_code"]])
+                elif user_choice == "project":
+                    url = reverse(
+                        'pootle-project-browse',
+                        args=[kwargs["project_code"], '', ''])
+                if url:
+                    response = redirect(url)
+                    response.delete_cookie('user-choice')
+                    return response
+            raise e
+        return f(self, request, *args, **kwargs)
+    return method_wrapper
+
+
+class TPMixin(object):
+
+    @redirect_to_tp_on_404
+    @set_permissions
+    @requires_permission("view")
+    def dispatch(self, request, *args, **kwargs):
+        # get funky with the request 8/
+        return super(TPMixin, self).dispatch(request, *args, **kwargs)
+
+    @property
+    def ctx_path(self):
+        return self.tp.pootle_path
+
+    @property
+    def resource_path(self):
+        return self.object.pootle_path.replace(self.ctx_path, "")
+
+    @property
+    def dir_path(self):
+        return self.resource_path
+
+    @cached_property
+    def tp(self):
+        return self.object.translation_project
+
+    @cached_property
+    def project(self):
+        return self.tp.project
+
+    @cached_property
+    def language(self):
+        return self.tp.language
+
+    @cached_property
+    def sidebar_announcements(self):
+        return get_sidebar_announcements_context(
+            self.request,
+            (self.project, self.language, self.tp))
+
+
+class TPDirectoryMixin(TPMixin):
+    model = Directory
+    browse_url_path = "pootle-tp-browse"
+    export_url_path = "pootle-tp-export"
+    translate_url_path = "pootle-tp-translate"
+
+    @lru_cache()
+    def get_object(self):
+        return get_object_or_404(
+            Directory.objects.all(),
+            pootle_path=self.path)
+
+    @property
+    def url_kwargs(self):
+        return {
+            "language_code": self.language.code,
+            "project_code": self.project.code,
+            "dir_path": self.dir_path}
+
+
+class TPStoreMixin(TPMixin):
+    model = Store
+    browse_url_path = "pootle-tp-store-browse"
+    export_url_path = "pootle-tp-store-export"
+    translate_url_path = "pootle-tp-store-translate"
+    is_store = True
+
+    @property
+    def permission_context(self):
+        return self.get_object().parent
```

Comment:
@phlax, I found that we call for get_object() more than one time for the same object. Should we keep it as attribute in this class?